services/src/main/java/org/keycloak/services/resources/admin/RealmsAdminResource.java

```diff
@@ -93,14 +94,26 @@ public RealmsAdminResource(AdminAuth auth, TokenManager tokenManager) {
     @GET
     @NoCache
     @Produces(MediaType.APPLICATION_JSON)
-    public Stream<RealmRepresentation> getRealms() {
+    public Stream<RealmRepresentation> getRealms(@QueryParam("basic") String basic) {
         Stream<RealmRepresentation> realms = session.realms().getRealmsStream()
-                .map(this::toRealmRep)
+                .map(realm -> toRealmRep(realm, Boolean.valueOf(basic)))
                 .filter(Objects::nonNull);
         return throwIfEmpty(realms, new ForbiddenException());
     }
 
     protected RealmRepresentation toRealmRep(RealmModel realm) {
+        return toRealmRep(realm, false);
+    }
+
+    protected RealmRepresentation toRealmRep(RealmModel realm, boolean basic) {
+        if (basic) {
+            RealmRepresentation rep = new RealmRepresentation();
+
+            rep.setId(realm.getId());
+            rep.setRealm(realm.getName());
+
+            return rep;
```

Comment:
@mhajas Thanks for all the details about the issue. Looks like I missed the checks for filtering realms. It should be fixed now.

I'm not sure why you did not get the menu though. You were supposed to get that only after creating a new realm. Which now is fixed.

I also tried to make some improvements when accessing a particular realm when using admins other than master admin.

I agree we should run UI tests. Please let me know if you spot anything else because I know our UI tests do not provide full coverage.

Reply:
@mhajas Btw, if you are testing performance as well, you should also consider increasing the number of entries in both realms and users cache.

My latest changes should reduce the time taken to get a token to ms, instead of seconds.

I can share with you the database dump I'm using for tests.

Reply:
Thank you @pedroigor. It seems all issues I mentioned are fixed now. I played with it a little more and there is one more change in behaviour. When you start fresh keycloak with an admin user, after you navigate to admin console and log in, admin console is always asking what realm you want to manage (at `#/realms`), even when there is only one, while previously it opened `#/realms/master`. Is this change intentional?

Reply:
@mhajas It should be fixed now.